spark/src/test/scala/org/apache/spark/sql/delta/typewidening/TypeWideningConstraintsSuite.scala

```diff
@@ -133,25 +133,58 @@ trait TypeWideningConstraintsTests { self: QueryTest with TypeWideningTestMixin
           },
           errorClass = "DELTA_CONSTRAINT_DATA_TYPE_MISMATCH",
           parameters = Map(
-            "columnName" -> "a",
-            "columnType" -> "STRUCT<x: TINYINT, y: TINYINT>",
-            "dataType" -> "STRUCT<x: INT, y: TINYINT>",
+            "columnName" -> "a.x",
+            "columnType" -> "TINYINT",
+            "dataType" -> "INT",
             "constraints" -> "delta.constraints.ck -> hash ( a . x ) > 0"
-        ))
+          )
+        )
+
+        // changing the type of struct field `a.y` when it's not
+        // the field referenced by the CHECK constraint is allowed.
+        sql("INSERT INTO t (a) VALUES (named_struct('x', CAST(2 AS byte), 'y', 500))")
+        checkAnswer(sql("SELECT hash(a.x) FROM t"), Seq(Row(1765031574), Row(1765031574)))
+      }
+    }
+  }
 
-        // We're currently too strict and reject changing the type of struct field a.y even though
-        // it's not the field referenced by the CHECK constraint.
+  test("check constraint on nested field with complex type evolution") {
+    withTable("t") {
+      sql("CREATE TABLE t (a struct<x: struct<z: byte, h: byte>, y: byte>) USING DELTA")
+      sql("ALTER TABLE t ADD CONSTRAINT ck CHECK (hash(a.x.z) > 0)")
+      sql("INSERT INTO t (a) VALUES (named_struct('x', named_struct('z', 2, 'h', 3), 'y', 4))")
+      checkAnswer(sql("SELECT hash(a.x.z) FROM t"), Row(1765031574))
+
+      withSQLConf(DeltaSQLConf.DELTA_SCHEMA_AUTO_MIGRATE.key -> "true") {
         checkError(
           exception = intercept[DeltaAnalysisException] {
-            sql("INSERT INTO t (a) VALUES (named_struct('x', CAST(2 AS byte), 'y', 500))")
+            sql(
+              s"""
+                 | INSERT INTO t (a) VALUES (
+                 |   named_struct('x', named_struct('z', 200, 'h', 3), 'y', 4)
+                 | )
+                 |""".stripMargin
+            )
           },
           errorClass = "DELTA_CONSTRAINT_DATA_TYPE_MISMATCH",
           parameters = Map(
-            "columnName" -> "a",
-            "columnType" -> "STRUCT<x: TINYINT, y: TINYINT>",
-            "dataType" -> "STRUCT<x: TINYINT, y: INT>",
-            "constraints" -> "delta.constraints.ck -> hash ( a . x ) > 0"
-        ))
+            "columnName" -> "a.x.z",
+            "columnType" -> "TINYINT",
+            "dataType" -> "INT",
+            "constraints" -> "delta.constraints.ck -> hash ( a . x . z ) > 0"
+          )
+        )
+
+        // changing the type of struct field `a.y` and `a.x.h` when it's not
+        // the field referenced by the CHECK constraint is allowed.
+        sql(
+          """
+            | INSERT INTO t (a) VALUES (
+            |   named_struct('x', named_struct('z', CAST(2 AS BYTE), 'h', 2002), 'y', 1030)
+            | )
+            |""".stripMargin
+        )
+        checkAnswer(sql("SELECT hash(a.x.z) FROM t"), Seq(Row(1765031574), Row(1765031574)))
```

Comment:
Can you add one test with e.g. `a struct<arr: array<struct<x, y>>>` with a constraint defined on `a.arr[0].x` and changing the type of a.element.x and a.element.x both fail?

Reply:
yes, I did attempt to add the corresponding tests; the issue is that, for both versions of `checkDependentConstraints`, we are unable to block operations involving `constraints` or `generated columns` related to `a.arr[0].x`, i.e., the operations pass without exceptions being thrown.

this seems to be more of a bug rather than something that should be addressed in this PR - I’m considering either adding the tests with a `// FIXME: ...` comment to highlight the issue, or simply skipping this scenario for now.

Reply:
I created a github issue to track this: https://github.com/delta-io/delta/issues/3634, this is indeed a bug. I expect that use case of applying a check constraint or generated column on a specific array  or map element should be fairly rare.